### DISCUSSION TITLE
Python 3.6 support?

### DISCUSSION THREAD

**Discussion by fsvenson** [piccolo-orm/piccolo#215] [Q&A] [answered]
I'm  just curious if piccolo has a hard dependency on python 3.7 or if 3.6 could be supported? Completely understandable if not, seeing as 3.6 is getting deprecated soon. Hopefully, we'll get around to upgrading our old CentOS 7 environment eventually...

**Comment by dantownsend** [comment, marked as answer]
I think the only things which lock it to Python 3.7 are dataclasses, but it seems like there's a backport available:

https://pypi.org/project/dataclasses/

And contextvars, which also seems to have a backport available.

https://pypi.org/project/contextvars/

So it might be possible. If it's as simple as adding those packages as optional dependencies, then it might be worth doing. I'll create an issue for it.

**Reply by fsvenson** [reply]
Thanks for the reply. Out of curiousity I tried myself to see whether I could get it to run, but it looks like the main issue is "from __future__ import annotations" and the use of forward-referencing. From what I can tell annotations won't work with Python 3.6.

**Reply by dantownsend** [reply]
Good point, I forgot about `from __future__ import annotations`. I haven't been able to find a work around for that unfortunately.

**Reply by aminalaee** [reply]
@dantownsend Can we annotate using strings like `"Class"` for all versions?

**Reply by dantownsend** [reply]
@aminalaee You're right, that is an option. It would be a pretty big change though. There's also this:

https://github.com/asottile/future-annotations

We would have to add text to the top of every file, which is also quite a big change.

It looks like Python 3.6 will be end of life at the end of the year, and it doesn't seem like there's a low friction way to backport it.

https://endoflife.date/python

**Reply by aminalaee** [reply]
@dantownsend Agreed. It's not worth it.
